listener: split listenerfilterfactorycontext and networkfilterfactorycontext

lambdai opened the pull request:
Description:
Extract from #9773 
The ultimate goal is to update network filter chains without destructing all the network filter chains in previous listener config.
Splitting lifetime of listener filter and network filter, so that we can reuse the references to the existing listener filters in old listener (will do in follow up PR or see #9773) 
The next PR is to arm the connection handler with in place filter chain update.

Also eliminate the risk of triggering warmed up listener when destroying ListenerImpl. 

Risk Level: LOW since the life time is equivalent with and without this PR
Testing: Fixed existing tests
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

lambdai commented:
/assign @mattklein123 since you have the contexts on #9773 

I admit this is not the smallest change I can extract but the filter chain manager and the drain manager change is necessary in the nest steps

lambdai commented:
/azp run envoy-macos

mattklein123 commented on include/envoy/event/deferred_task.h:
```diff
@@ -0,0 +1,27 @@
+#pragma once
```
I don't think this is used in this PR, please remove.

lambdai replied:
Will do

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -335,6 +336,14 @@ class FilterChain {
 using FilterChainSharedPtr = std::shared_ptr<FilterChain>;
 
+/**
+ * The filter chains could be marked as draining.
```
```suggestion
 * A filter chain that can be drained.
```

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -335,6 +336,14 @@ class FilterChain {
+ */
+class DrainableFilterChain : public FilterChain {
+public:
+  virtual void setDrainClose() PURE;
```
What does this actually do? Start the drain sequence? `startDrainSequence()` or similar?

lambdai replied:
More like a pure ` DrainManagerImpl::draining_ = true`. Once it is set, all the network filters(AFAIK only HCM) could poll the factory context for the state that the filter chain is going to be destructed.

mattklein123 replied:
OK then lets call this `startDraining()` or just `drain()`

lambdai replied:
will fix

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -448,5 +457,14 @@ class FilterChainFactory {
                                             UdpReadFilterCallbacks& callbacks) PURE;
 };
 
+class DrainingFilterChains {
```
I don't see this used in this PR. Please remove.

lambdai replied:
Agree... Removing it and create the next PR soon

mattklein123 commented on include/envoy/server/filter_config.h:
```diff
@@ -116,6 +117,11 @@ class ServerFactoryContext : public virtual CommonFactoryContext {
    */
   virtual Grpc::Context& grpcContext() PURE;
+
+  /** @return DrainManager& drain manager which has life no shorter than the life of the current
```
This is the server context so by definition isn't the lifetime of the entire server?

lambdai replied:
Theoretically yes. We seldom violates it... e.g. the initManager()

lambdai replied:
Let me think about it if I could done better

lambdai replied:
IIRC add this because I need to access the server drain manager. I didn't expect derived class returns a shorter life. Will update the comment.

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -71,8 +77,37 @@ class FilterChainFactoryContextImpl : public Configuration::FilterChainFactoryCo
   Configuration::TransportSocketFactoryContext& getTransportSocketFactoryContext() const override;
   Stats::Scope& listenerScope() override;
 
+  void setDraining() { is_draining_.store(true, std::memory_order_release); }
```
I think we can avoid the explicit memory order complexity for this. Let's just go with `is_draining_ = true;`

lambdai replied:
I think it's a straightforward publish/subscribe semantic so that I choose acquire/release
But I agree it's not a critical code path.

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -71,8 +77,37 @@ class FilterChainFactoryContextImpl : public Configuration::FilterChainFactoryCo
 private:
   Configuration::FactoryContext& parent_context_;
+  Init::Manager& init_manager_;
+  // atomic_flag is guaranteed to be lock free but there is no light weight load().
```
nit: del comment (we don't need to document the atomic behavior, we don't do this anywhere else).

mattklein123 commented on source/server/filter_chain_manager_impl.cc:
```diff
@@ -26,20 +26,17 @@ Network::Address::InstanceConstSharedPtr fakeAddress() {
 bool FilterChainFactoryContextImpl::drainClose() const {
-  // TODO(lambdai): will provide individual value for each filter chain context.
-  return parent_context_.drainDecision().drainClose();
+  return is_draining_.load(std::memory_order_acquire) ||
```
per below just `is_draining_` (we don't need the explicit load)

mattklein123 commented on source/server/filter_chain_manager_impl.cc:
```diff
@@ -144,6 +148,7 @@ void FilterChainManagerImpl::addFilterChain(
     FilterChainFactoryContextCreator& context_creator) {
   std::unordered_set<envoy::config::listener::v3::FilterChainMatch, MessageUtil, MessageUtil>
       filter_chains;
+  int new_filter_chain_size = 0;
```
s/int/uint32_t

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -122,18 +157,26 @@ class FactoryContextImpl : public Configuration::FactoryContext {
                                Logger::Loggable<Logger::Id::config> {
 public:
+  using FcContextMap =
+      std::unordered_map<envoy::config::listener::v3::FilterChain,
```
absl::flat_hash_map

lambdai replied:
Fixing.. thanks!

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -234,34 +282,27 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
-  }
+  // Reference to the previous generation of filter chain manager to share the filter chains.
+  // Caution: the pointer is legit only during warm up.
+  // TODO(lambdai): safer usage
```
What does this mean? The origin_ concept seems pretty strange. Why does it need to be stored?

lambdai replied:
I considered pointer because the original could be none. Let me see if I can remove

lambdai replied:
I probably missed the context here due to the extraction. 
The origin filter chain manager need to be used to co-own the shared filter chains. It's not a pure transfer because the newer filter chain manager could be destroyed when another update replace the warming one.
The plan of safe usage is something like adding state_ to FilterChainManagerImpl, once the build phase is finished the origin cannot be accessed. Since the state transition is straightforward, I am not sure the plan is a overkill.

mattklein123 replied:
Then please remove the TODO and make the comment more complete per what you said here.

mattklein123 commented on source/server/listener_impl.h:
```diff
@@ -74,12 +76,130 @@ class ListenSocketFactoryImpl : public Network::ListenSocketFactory,
 //                     initializing all listeners after workers are started.
 
+/**
+ * The immutable factory context during the intelligent listener update. The continuous intelligent
```
What is an intelligent listener? Wrong comment?

lambdai replied:
The comment is back from future :)

Fixing

mattklein123 commented on source/server/listener_impl.cc:
```diff
@@ -153,13 +223,19 @@ ListenerImpl::ListenerImpl(const envoy::config::listener::v3::Listener& config,
+                            [this]() { dynamic_init_manager_->initialize(local_init_watcher_); }),
+      dynamic_init_manager_(std::make_unique<Init::ManagerImpl>(
+          fmt::format("Listener-local-init-manager {} {}", name, hash))),
+      // local_drain_manager_(parent.factory_.createDrainManager(config.drain_type())),
```
?

lambdai replied:
removed

mattklein123 commented on source/server/listener_impl.h:
```diff
@@ -74,12 +76,130 @@ class ListenSocketFactoryImpl : public Network::ListenSocketFactory,
+
+class ListenerImpl;
+// TODO(lambdai): Strip the interface since ListenerFactoryContext only need to support
+// ListenerFilterChain creation.
```
Please do this now. It's very hard to understand the new flow of classes/interfaces here. If you need to do an intermediate PR let's do that.

lambdai replied:
@mattklein123  Sorry I propose delay this strip. Previously ListenerFactoryContext supports both network filter chain creation and listener filter chain creation. 
So ListenerFactoryContext must be fat enough to cover both.
Obviously when we split the creation the two interface could be smaller but it take time to see if the filters in envoy repo and in 3rd party repo could survive. I assume we need a grace period to announce (even simply through issues) and to deprecate.

mattklein123 replied:
> Obviously when we split the creation the two interface could be smaller but it take time to see if the filters in envoy repo and in 3rd party repo could survive. I assume we need a grace period to announce (even simply through issues) and to deprecate.

We don't promise API compatibility so this can be done now, or please make it more clear in the TODO what needs to be done. The comment is not thorough enough to understand.

mattklein123 commented on source/server/listener_manager_impl.cc:
```diff
@@ -558,6 +558,10 @@ void ListenerManagerImpl::drainListener(ListenerImplPtr&& listener) {
+          // TODO(lambdai): Resolve race condition. The below refcount dec is too early. The active
+          // listener at worker thread could be destroyed before active connection.
+          // As a consequence, the listener at master thread and the connection at worker thread are
+          // destroyed concurrently.
```
I don't follow the race condition. Clarify?

lambdai replied:
Sorry for the unclear. I think I have an example in original PR.
Consider the below events in global sequence order
1. master thread: calling drainListener
2. work thread: *deferred delete* the active connection
3. work thread: post to master that the drain is done
4. master thread: erase the listener
5. worker thread: execute destroying connection when the shared listener config is destroyed at step 4 (could be worse such as access the connection because connection is not yet started to deleted).

The race condition is introduced because 3 occurs too early. My solution is to defer schedule the callback posting to master thread, by introducing DeferTaskUtil. So that 5 should always happen before 3.

mattklein123 replied:
OK we can cover this in a later PR. Until then please update the comment to be more complete as you have written here.

mattklein123 commented on source/server/listener_manager_impl.cc:
```diff
@@ -832,11 +838,17 @@ std::unique_ptr<Network::FilterChain> ListenerFilterChainFactoryBuilder::buildFi
+                                            *message, factory_context_, std::move(server_names)),
+                                        listener_component_factory_.createNetworkFilterFactoryList(
+                                            filter_chain.filters(), *filter_chain_factory_context));
+  // TODO(lambdai): remove this ugly downcast
```
Why is this needed?

lambdai replied:
Ooops, i think it's not required. Let me check

lambdai replied:
I don't have an immediate solution on fixing this...
The background is that we only have one implementation at downstream(listener, filter chain manager, network filter chain) but we do have network filter chain at upstream. I am adding downstream filter chains the drainable functionality per downstream filter chain context so it rely on implementation not the filter chain context common interface.

Meanwhile it's safe at this moment because ListenerManager knows exactly what it is doing

mattklein123 replied:
This is fragile let's please fix this now. You might need to somehow split the upstream/downstream filter chain interfaces. If so, please do that in a separate PR first.

lambdai commented:
bazel tsan failing because the integration test is accessing listener manager from different thread but listener manager is not thread safe. If the failure rate is low it should be fine

lambdai commented:
test should be fixed in https://github.com/envoyproxy/envoy/pull/10523

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <list>
```
nit: is this needed?

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -71,8 +77,36 @@ class FilterChainFactoryContextImpl : public Configuration::FilterChainFactoryCo
+  }
+
+  void startDraining() override { factory_context_->startDraining(); }
+  std::unique_ptr<FilterChainFactoryContextImpl> factory_context_;
```
nit: does this need to be public? Can you make this private and accessed via accessor?

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -145,6 +187,11 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
       FilterChainFactoryBuilder& b, FilterChainFactoryContextCreator& context_creator);
   static bool isWildcardServerName(const std::string& name);
 
+  // TODO(lambdai): move to private
```
Please just do this now, similar to my other comment.

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -234,34 +281,27 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
   // and application protocols, using structures defined above.
   DestinationPortsMap destination_ports_map_;
   const Network::Address::InstanceConstSharedPtr address_;
+  // This is the reference to a factory context where all the generations of listener share.
```
```suggestion
  // This is the reference to a factory context which all the generations of listener share.
```

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -234,34 +281,27 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
-  const Network::TransportSocketFactoryPtr transport_socket_factory_;
-  const std::vector<Network::FilterFactoryCb> filters_factory_;
+  // For FilterChainFactoryContextCreator
+  // init manager owned by the corresponding listener. The reference is legit when building the
```
s/legit/valid

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -234,34 +281,27 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
-    return filters_factory_;
-  }
+  // Reference to the previous generation of filter chain manager to share the filter chains.
+  // Caution: the pointer is legit only during warm up.
```
s/legit/valid, same elsewhere that I missed.

mattklein123 commented on source/server/listener_impl.h:
```diff
@@ -74,12 +76,128 @@ class ListenSocketFactoryImpl : public Network::ListenSocketFactory,
+ * The immutable factory context could be used by listener filter factory contexts and network
+ * filter factory contexts.
+ */
+class ListenerFactoryContextBaseImpl final : public Configuration::FactoryContext,
```
This is per listener right? Let's call this `PerListenerFactoryContextImpl`. 

The one below is per filter chain? Let's call that `PerFilterChainFactoryContextImpl`.

lambdai replied:
This `ListenerFactoryContextBaseImpl` is the common part of below per listener and per filter chain context. It wraps the `Server` instance and a `ListenerImpl` instance.

The below `ListenerFactoryContextImpl` is per listener, I renamed to `PerListenerFactoryContextImpl`
The `FilterChainFactoryContextImpl` is per filter chain and I renamed to `PerFilterChainFactoryContextImpl`

mattklein123 commented on source/server/listener_impl.cc:
```diff
@@ -123,15 +123,85 @@ Network::SocketSharedPtr ListenSocketFactoryImpl::getListenSocket() {
+    const envoy::config::listener::v3::Listener& config, DrainManagerPtr drain_manager)
+    : server_(server), metadata_(config.metadata()), direction_(config.traffic_direction()),
+      global_scope_(server.stats().createScope("")),
+      // Not ideal
```
? Comment not actionable.

lambdai replied:
I remember I resolve addresses multiple times... NBD. Deleted the comment and we will rediscover if it matters

mattklein123 commented on source/server/listener_impl.cc:
```diff
@@ -123,15 +123,85 @@ Network::SocketSharedPtr ListenSocketFactoryImpl::getListenSocket() {
+  return server_.threadLocal();
+}
+Admin& ListenerFactoryContextBaseImpl::admin() { return server_.admin(); }
+// TODO(lambdai): consider moving away from factory context
```
Is this a duplicate TODO? If so please remove or make it more clear.

lambdai replied:
Right. This is one of the accessor in listener filter interface: no known code is using it but obviously it is in the listener api for some reason.

mattklein123 commented on source/server/listener_impl.cc:
```diff
@@ -427,9 +525,11 @@ ListenerImpl::~ListenerImpl() {
     // parent's initManager to get ready().
     listener_init_target_.ready();
   }
+  // destruct immediately to prevent other member destruction triggered the ready().
```
nit: Start comment with capital end with period, please check the rest of the PR.

mattklein123 replied:
Also, why is this needed? This is in the destructor. Can you just reorder the members if that matters?

lambdai replied:
Will do. Add a TODO here to avoid too many things in this PR

mattklein123 commented on test/integration/integration.cc:
```diff
@@ -409,10 +409,25 @@ void BaseIntegrationTest::setUpstreamAddress(
 }
 
+namespace {
+bool read_flag(bool* flag) { return *flag; }
```
We can merge master once the other PR is reviewed and merged.

lambdai replied:
Of course. Early merge it here only to prove this PR is well tested.

mattklein123 commented on test/mocks/network/mocks.h:
```diff
@@ -468,6 +469,5 @@ class MockConnectionBalancer : public ConnectionBalancer {
   MOCK_METHOD(BalancedConnectionHandler&, pickTargetHandler,
               (BalancedConnectionHandler & current_handler));
 };
-
```
revert extraneous changes here/elsewhere.

lambdai commented:
I removed most of the TODOs. 

Two major TODOs are left:
1. Race condition due to deferred destruction. As we agreed, it's not a regression and I will fix afterwards. 
2. Shim the interface FactoryContext for network filter and listener filter. It's an arguable what scope a `network filter` should refer a `listener`, comparing with a `listener filter`. Prior to this PR, there is no need to discuss because both network filter and listener filters refers the same factory context instance. For compatibility I grants the same ability to network filter as it was prior to this PR. I will take several iterations to gradually tailor the dependencies of network filter factory context.

lambdai commented:
early introduce another fix https://github.com/envoyproxy/envoy/pull/10579 to fix the surfacing flaky h1 fuzz persistent tests

mattklein123 commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -234,34 +282,36 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
-                  std::vector<Network::FilterFactoryCb>&& filters_factory)
-      : transport_socket_factory_(std::move(transport_socket_factory)),
-        filters_factory_(std::move(filters_factory)) {}
+  bool add_filter_chain_done_{false};
```
Instead of an additional boolean I would use absl::optional if you need to differentiate between null and valid.

lambdai replied:
Agree!  Fixed

mattklein123 commented on source/server/listener_impl.cc:
```diff
@@ -123,15 +123,85 @@ Network::SocketSharedPtr ListenSocketFactoryImpl::getListenSocket() {
+  return server_.threadLocal();
+}
+Admin& ListenerFactoryContextBaseImpl::admin() { return server_.admin(); }
+// TODO(lambdai): Consider moving away from factory context since listener filter and network filter
+// have independent factory context.
```
It's unclear where this TODO actually applies or what it means. Can you remove this or put this next to class docs somewhere?

lambdai replied:
Moved there and list listenerMetadata as an example

mattklein123 commented on source/server/listener_impl.cc:
```diff
@@ -427,9 +526,13 @@ ListenerImpl::~ListenerImpl() {
     listener_init_target_.ready();
   }
+
+  // TODO(lambdai): Reorder the members to simplify the dtor.
```
Please just do this now. It should be a trivial change.

lambdai replied:
done

lambdai commented:
merging master and resolving conflict with my other PR
Seems I hit another static initialized string issue :(

lambdai commented:
This is the closest situation to all check passing during these days. @mattklein123 I addressed all your comments. Could you take another (hopefully last) look? Thanks!

lambdai commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -234,34 +282,34 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
                                     const Network::ConnectionSocket& socket) const;
 
+  const FilterChainManagerImpl* getOriginFilterChainManager() {
+    ASSERT(origin_.has_value());
```
I thought we prefer prefer these precondition-ish ASSERT since the assertion failure is more friendly in debugging mode.
Sure, I will update in the next PR. Thank you for allowing me amending in follow up!

mergeconflict commented on source/server/filter_chain_factory_context_callback.h:
```diff
@@ -21,7 +21,7 @@ class FilterChainFactoryContextCreator {
    * Generate the filter chain factory context from proto. Note the caller does not own the filter
    * chain context.
```
Quick drive-by comment: IIUC, the caller does own the filter chain context now, is that right?
